### PR TITLE
Function for setting age-structured outputs.

### PR DESCRIPTION
**Pull request by RJSheppard**
Sets parameters that output age-structured incidence, prevalence and immunity levels. Discrete age groups are set using a list, where vectors within the list result in contiguous age group rendering.

(I also corrected sp. of prevelance -> prevalence :) )

### CODE REVIEW

File: R/output_parameters.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#' @title Parameterise age grouped output rendering
```
Comment (pwinskill) [inline R/output_parameters.R:1]: I found this a little tricky to parse (although I think it all works!).

I was wondering if we can avoid having to input lists, and (maybe) make the process inthe function a bit simpler. Does the following work in the correct way? I haven't tested it!

``` 
#' @title Parameterise age grouped output rendering
#'
#' @details this function produces discrete and contiguous age groups, inclusive of the lower
#' age limit and exclusive of the upper age limit: e.g., c(0, 10, 100, 200, 250) will produce
#' three age groups: 0-9, 10-99 and 200-249 in days.
#' @param parameters the model parameters
#' @param age_group age breaks for population size outputs; default = NULL
#' @param incidence age breaks for incidence outputs (D+Tr+A); default = NULL
#' @param clinical_incidence age breaks for clinical incidence outputs (symptomatic); default = c(0, 1825)
#' @param severe_incidence age breaks for severe incidence outputs; default = NULL
#' @param prevalence age breaks for clinical prevalence outputs (pcr and lm detectable infections); default = c(730, 3650)
#' @param ica age breaks for average acquired clinical immunity; default = NULL
#' @param icm age breaks for average maternal clinical immunity; default = NULL
#' @param iva age breaks for average acquired severe immunity; default = NULL
#' @param ivm age breaks for average maternal severe immunity; default = NULL
#' @param id age breaks for average immunity to detectability; default = NULL
#' @param ib age breaks for average blood immunity; default = NULL
#' @export
#'
set_epi_outputs <- function(parameters,
                            age_group = c(0, 10, 100, 200, 250),
                            incidence = NULL,
                            clinical_incidence = c(0, 1825),
                            severe_incidence = NULL,
                            prevalence = c(730, 3650),
                            ica = NULL,
                            icm = NULL,
                            iva = NULL,
                            ivm = NULL,
                            id = NULL,
                            ib = NULL
){
  
  input <- list(
    age_group = age_group,
    incidence = incidence,
    clinical_incidence = clinical_incidence,
    severe_incidence = severe_incidence,
    prevalence = prevalence,
    ica = ica,
    icm = icm,
    iva = iva,
    ivm = ivm,
    id = id,
    ib = ib
  )
  input <- input[!sapply(input, is.null)]
  
  for(i in seq_along(input)){
    name <- names(input)[i]
    ages <- input[[i]]
    min_ages <- ages[-length(ages)]
    max_ages <- ages[-1] - 1
    parameters[[paste0(name, "_rendering_min_ages")]] <- min_ages
    parameters[[paste0(name, "_rendering_max_ages")]] <- max_ages
  }
  
  return(parameters)
}
```

Comment (RJSheppard) [reply R/output_parameters.R:1]: I was trying to capture discrete age blocks with the lists, e.g., if you want to output 1-100 and 200-300, but not 101-199. I don't think this code would do that, but maybe that's not so important for it to do.

File: R/output_parameters.R [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+){
+  
+  parent_formals <- names(formals())
+  parent_formals <- parent_formals[which(parent_formals != "parameters")]
+  outputs <- parent_formals[!unlist(lapply(parent_formals, function(x){is.null(get(x))}))]
+  
+  for (output in outputs) {
+    if(!is.list(get(output))){stop("Each age input must be a list of vectors")}
+    parameters[[paste0(output, "_rendering_min_ages")]] <- unlist(lapply(get(output), function(x){x[-length(x)]}))
+    parameters[[paste0(output, "_rendering_max_ages")]] <- unlist(lapply(get(output), function(x){x[-1]-1}))
```
Comment (giovannic) [inline R/output_parameters.R:41]: I agree with Pete that it's a little terse. You could perhaps make it more readable with names for the functions `function(x){x[-length(x)]}` and `function(x){x[-1]-1}`, and/or unwrapping an unlist/lapply into a for loop.

**Review by giovannic — CHANGES_REQUESTED** [review]

File: R/output_parameters.R (hunk 2 above)
Comment (RJSheppard) [reply R/output_parameters.R:41]: Sure thing - happy to make this explicit.

File: R/output_parameters.R [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                            iva = NULL,
+                            ivm = NULL,
+                            id = NULL,
+                            ib = NULL
+){
+  
+  parent_formals <- names(formals())
+  parent_formals <- parent_formals[which(parent_formals != "parameters")]
+  outputs <- parent_formals[!unlist(lapply(parent_formals, function(x){is.null(get(x))}))]
+
```
[the remark this answers is not in the document]
Comment (RJSheppard) [reply R/output_parameters.R:37]: Sorry - would you mind explaining this a bit more? Are you saying to just check that the parameters appear in the parameter list after this function is used? I have a test that should check for this.

**Comment by RJSheppard** [comment]
Was there anything extra on this branch that either of you wanted changing (@pwinskill @giovannic)?

**Review by pwinskill — APPROVED** [review]
👍

**Pull request merged**